src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,562 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.marshal;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.cassandra.cql3.CQL3Type;
+import org.apache.cassandra.cql3.Term;
+import org.apache.cassandra.cql3.Vectors;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.serializers.TypeSerializer;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.JsonUtils;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+public final class VectorType<T> extends AbstractType<List<T>>
```

Comment:
My understanding was that this would be `public final class VectorType<T> extends AbstractType<T[]>`. This fits better with the serialization and expectations on the driver. 

I have tested this and have a working prototype that works this way. There was one stumbling block here and that is that the `TypeSerializer.getType` cannot be made to work with a generic type. Luckily this method is not actually used anywhere so I suggest that we get rid of it.

Reply:
> This fits better with the serialization and expectations on the driver.

from the drivers point of view, `List` vs java array doesn't matter... that's a internal detail for serializer on server; byte format isn't impacted.

Issue with `T[]` is that we would then need to have `Class` as a field so we can build such generic arrays... and adds a lot more complication.

Reply:
Agreed. I think this answers some of my questions above as well. We can leave the driver details to later.